keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1475,225 +1382,400 @@ prevented. Deleting the entire cluster also leaves resources allocated outside
+    // access is requested. Admin access is granted also for
+    // devices which are in use.
+    //
+    // The default if `all` and `count` are unset is to allocate one device.
+    //
+    // +optional
+    All bool
 
-    // DriverRequests describes all resources that are needed for the
-    // allocated claim. A single claim may use resources coming from
-    // different drivers. For each driver, this array has at most one
-    // entry which then may have one or more per-driver requests.
+    // A fixed number of devices which match the same selectors.
     //
-    // May be empty, in which case the claim can always be allocated.
-    DriverRequests []DriverRequests
+    // The default if `all` and `count` are unset is to allocate one device.
+    //
+    // +optional
+    Count *int64
 }
 
-// DriverRequests describes all resources that are needed from one particular driver.
-type DriverRequests struct {
-    // DriverName is the name used by the DRA driver kubelet plugin.
-    DriverName string
+// Besides the request name slice, constraint must have one and only one field set.
+type Constraint struct {
+    // The constraint applies to devices in these requests. A single entry is okay
+    // and used when that request is for multiple devices.
+    //
+    // If empty, the constrain applies to all devices in the claim.
+    //
+    // +optional
+    RequestNames []string
 
-    // VendorParameters are arbitrary setup parameters for all requests of the
-    // claim. They are ignored while allocating the claim.
-    VendorParameters runtime.Object
+    // If set, then the devices must have this attribute and its value must be the same.
+    //
+    // For example, if you specified "numa.dra.example.com" (a hypothetical example!),
+    // then only devices in the same NUMA node will be chosen.
+    //
+    // +optional
+    // +listType=atomic
+    MatchAttribute *string
 
-    // Requests describes all resources that are needed from the driver.
-    Requests []ResourceRequest
+    // Future extension, not part of the current design:
+    // A CEL expression which compares different devices and returns
+    // true if they match.
+    //
+    // Because it would be part of a one-of, old schedulers will not
+    // accidentally ignore this additional, for them unknown match
+    // criteria.
+    //
+    // matcher string
 }
 
-// ResourceRequest is a request for resources from one particular driver.
-type ResourceRequest struct {
-    // VendorParameters are arbitrary setup parameters for the requested
-    // resource. They are ignored while allocating a claim.
-    VendorParameters runtime.Object
+type ClaimConfigurationParameters struct {
+    // The configuration applies to devices in these requests.
+    //
+    // If empty, the configuration applies to all devices in the claim.
+    //
+    // +optional
+    RequestNames []string
 
-    ResourceRequestModel
+    ConfigurationParameters // inline
```

Comment:
Again, why not put the fields here?

Reply:
In this case the reason is that `ConfigurationParameters` (now called just `Configuration`) gets embedded three times. If I keep it separate, I can write a single `validateConfiguration` function instead of having to copy-and-paste that code (Go generics don't support access to fields that are the same in different types).

I don't feel strongly about this.

Reply:
ACK